socrata/socrata_upload.py

```diff
@@ -0,0 +1,337 @@
+    )
+
+    tic = time.perf_counter()
+    count = 0
+
+    if not flag:
+        sql_query = build_query(
+            athena_asset=athena_asset,
+            row_identifier=row_identifier,
+        )
+
+        upload_args = {
+            "asset_id": asset_id,
+            "sql_query": sql_query,
+            "overwrite": overwrite,
+        }
+
+        if overwrite:
+            response = upload("put", **upload_args)
+        else:
+            response = upload("post", **upload_args)
+        print(response.content)
+
+    else:
+        if flag == "years":
+            sql_query = build_query(
+                athena_asset=athena_asset,
+                row_identifier=row_identifier,
+                years=years,
+            )
+
+        if flag == "both":
+            sql_query = build_query(
+                athena_asset=athena_asset,
+                row_identifier=row_identifier,
+                years=years,
+                township=by_township,
+            )
+
+        if overwrite:
+            for item in groups:
+                if flag == "both":
+                    upload_args = {
+                        "asset_id": "asset_id",
+                        "sql_query": "sql_query",
+                        "overwrite": "overwrite",
+                        "year": item[0],
+                        "township": item[1],
+                    }
+                elif flag == "years":
+                    upload_args = {
+                        "asset_id": asset_id,
+                        "sql_query": sql_query,
+                        "overwrite": overwrite,
+                        "year": item,
+                    }
+                if count == 0:
+                    response = upload("put", **upload_args)
+                else:
+                    response = upload("post", **upload_args)
+                print(response.content)
+                count = count + 1
+
+        else:
+            for item in groups:
+                if flag == "both":
+                    upload_args = {
+                        "asset_id": asset_id,
+                        "sql_query": sql_query,
+                        "overwrite": overwrite,
+                        "year": item[0],
+                        "township": item[1],
+                    }
+                elif flag == "years":
+                    upload_args = {
+                        "asset_id": asset_id,
+                        "sql_query": sql_query,
+                        "overwrite": overwrite,
+                        "year": item,
+                    }
+                response = upload("post", **upload_args)
+                print(response.content)
+
+    toc = time.perf_counter()
+    print(f"Total upload in {toc - tic:0.4f} seconds")
+
+
+socrata_upload(
+    socrata_asset=os.getenv("SOCRATA_ASSET"),
+    overwrite=os.getenv("OVERWRITE"),
+    years=os.getenv("YEARS"),
+    by_township=os.getenv("BY_TOWNSHIP"),
+)
```

Comment:
**[Thought, non-blocking]** One high-level thought that I have for you about the design of the functions in this module: Currently, there are three possible filter conditions (no filters; all filters; or year _but not_ township filter) and each one of the `build_query`, `generate_groups`, and `upload` functions are responsible for implementing a different solution for each condition. This works fine, but it makes the functions hard to read since each one has its own set of complex conditional branches, and if we ever needed to alter those conditions or add a new one I think we would find ourselves having to basically rewrite the whole module to accommodate the new or changed condition.

I think there's potential for a simpler design that consolidates all of the responsibility for evaluating the filters in the `generate_groups` function, which could return a list of groups that could then be operated on by each of the `build_query` and `upload` functions agnostic of the exact contents of the filters themselves. This is a bit speculative, so I would be happy to take a stab at a follow-up PR to demonstrate a refactor if you're interested but not sure what I mean. Let me know and we can hash it out! But I don't think it's necessary to get too deep into it while you're still trying to just get the functionality up and running.